### PR TITLE
Restructure and extend

### PR DESCRIPTION
**Pull request by rcannood**
## New features

* Added `project/sync-and-cache-s3`: Sync and cache an S3 bucket.

* Added `project/detect-changed-components`: Filter a json produced by `ns-list` such that
  only components whose resources have changed in comparison to the default branch
  are retained.

## Breaking changes

* `pro/build-nextflow-params`:
  - Renamed from `build-nextflow-params`.
  - Input `token` was renamed to `viash_pro_token`.

* `pro/build-nextflow-schemas`:
  - Renamed from `build-nextflow-schemas`.
  - Input `token` was renamed to `viash_pro_token`.

* `pro/generate-documentation-qmd`:
  - Renamed from `generate-documentation-qmd`.
  - Input `token` was renamed to `viash_pro_token`.

### CODE REVIEW

File: project/sync-and-cache-s3/action.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    - name: Create hash key
+      shell: bash
+      id: cache_hash
+      run: |
+        AWS_EC2_METADATA_DISABLED=true \
+          aws s3 ls \
+          ${{ inputs.s3_bucket }} \
+          --recursive \
+          --no-sign-request \
+          > bucket-contents.txt
```
[the remark this answers is not in the document]
Comment (rcannood) [reply project/sync-and-cache-s3/action.yml]: Done! I simply piped the output directly to md5sum so the temp file is no longer needed :)

File: project/detect-changed-components/action.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+          (.info.config | capture("^(?<dir>.*\/)").dir) as $dir |
+          ([.info.config] +
+            ([.functionality.resources[].path?] | map($dir + .)) +
+            ([.functionality.test_resources[].path?] | map($dir + .)))[]
+        ' <<< "$component")
+        
+        # check if resource is in the list of changed resources
+        for resource_rel_path in "${resources[@]}"; do
+          resource_project_path=$(realpath --relative-to="$GITHUB_WORKSPACE" "$resource_rel_path")
+          if [[ " ${changed_files[*]} " =~ " ${resource_project_path} " || "$GITHUB_REF" == "refs/heads/main" || "${{ contains(github.event.head_commit.message, 'ci force') }}" == "true" ]]; then
```
Comment (DriesSchaumont) [inline project/detect-changed-components/action.yml:56]: Move the 'ci force' check up

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]
ALGTM!

File: project/detect-changed-components/action.yml (hunk 2 above)
Comment (rcannood) [reply project/detect-changed-components/action.yml]: :+1: !

File: project/detect-changed-components/action.yml [hunk 3]
```diff
@@ -28,40 +28,48 @@ runs:
...
-      # create output array
-      declare -a output_array=()
+      # create output path
+      output_file=$(mktemp $RUNNER_TEMP/viash_ns_changed_files.XXXXXX.json)
+
+      if [[ "$GITHUB_REF" == "refs/heads/${{ github.event.repository.default_branch }}" || "${{ contains(github.event.head_commit.message, 'ci force') }}" == "true" ]]; then
+        cp "${{ input.input_file }}" "$output_file"
+      else
+        # create output array
+        declare -a output_array=()
```
Comment (DriesSchaumont) [inline project/detect-changed-components/action.yml:38]: perhaps do

```
if []; then
 ...
 exit 0;
fi
```

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]

File: project/sync-and-cache-s3/action.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply project/sync-and-cache-s3/action.yml]: Even better! 👍

File: project/detect-changed-components/action.yml (hunk 3 above)
Comment (rcannood) [reply project/detect-changed-components/action.yml]: Done!

**Review by DriesSchaumont — APPROVED** [review]

**Pull request merged**